datafusion/physical-plan/src/aggregates/group_values/column.rs

```diff
@@ -75,55 +176,317 @@ pub struct GroupValuesColumn {
+            }
+            current_row_equal_to_result |= equal_to_result;
+
+            // Look forward next one row to check if have checked all results
+            // of current row
+            let next_row = self
+                .vectorized_equal_to_row_indices
+                .get(idx + 1)
+                .unwrap_or(&usize::MAX);
+
+            // Have checked all results of current row, check the total result
+            if row != *next_row {
+                // Not equal to case, add `row` to `scalarized_indices`
+                if !current_row_equal_to_result {
+                    self.scalarized_indices.push(row);
+                }
+
+                // Init the total result for checking next row
+                current_row_equal_to_result = false;
+            }
+        }
+
+        self.vectorized_equal_to_results = equal_to_results;
+    }
+
+    fn scalarized_equal_to(
+        &self,
+        group_index_view: &GroupIndexView,
+        cols: &[ArrayRef],
+        row: usize,
+        groups: &mut Vec<usize>,
+    ) -> bool {
+        // Check if this row exists in `group_values`
+        fn check_row_equal(
+            array_row: &dyn GroupColumn,
+            lhs_row: usize,
+            array: &ArrayRef,
+            rhs_row: usize,
+        ) -> bool {
+            array_row.equal_to(lhs_row, array, rhs_row)
+        }
+
+        if group_index_view.is_non_inlined() {
+            let list_offset = group_index_view.value() as usize;
+            let group_index_list = &self.group_index_lists[list_offset];
+
+            for &group_idx in group_index_list {
+                let mut check_result = true;
+                for (i, group_val) in self.group_values.iter().enumerate() {
+                    if !check_row_equal(group_val.as_ref(), group_idx, &cols[i], row) {
+                        check_result = false;
+                        break;
+                    }
+                }
+
+                if check_result {
+                    groups[row] = group_idx;
+                    return true;
+                }
+            }
+
+            // All groups unmatched, return false result
+            false
+        } else {
+            let group_idx = group_index_view.value() as usize;
+            for (i, group_val) in self.group_values.iter().enumerate() {
+                if !check_row_equal(group_val.as_ref(), group_idx, &cols[i], row) {
+                    return false;
+                }
+            }
+
+            groups[row] = group_idx;
+            true
+        }
+    }
+
+    fn scalarized_intern(
+        &mut self,
+        cols: &[ArrayRef],
+        batch_hashes: &[u64],
+        groups: &mut Vec<usize>,
+    ) {
+        if self.scalarized_indices.is_empty() {
+            return;
+        }
+
+        let mut map = mem::take(&mut self.map);
+
+        for &row in &self.scalarized_indices {
+            let target_hash = batch_hashes[row];
```

Comment:
Since we need the `group_index_view` of a specific row, which we already have in `collect_vectorized_process_context`, we could store `group_index_view` in another Vec? By doing this, we avoid using an additional map lookup.

```rust
                self.map.insert_accounted(
                    (target_hash, group_index_view),
                    |(hash, _)| *hash,
                    &mut self.map_size,
                );

                // Add row index to `vectorized_append_row_indices`
                self.vectorized_append_row_indices.push(row);

                // Something similar to this
                group_view_indexes[row] = group_index_view
```

Reply:
I found some corner cases exist, and seems we can't  store the `group_index_view`...

One case is `repeated rows` in `input batch`:
- all of them will be correspoding with same `group_index_view` in `collect_vectorized_process_context`.
- and assume that after we process the first of them, the rest stored `group_index_views` will be inconsistent.

The situation may be like:
```text
// Before process
row1 --> group_index_view (inlined, point to a group index)
row1 --> group_index_view (inlined, point to a group index)
row1 --> group_index_view (inlined, point to a group index)

// After process the first one
row1 --> group_index_view (uninline, point to group indices)
row1 --> group_index_view (inlined, inconsistent)
row1 --> group_index_view (inlined, inconsistent)
```

Reply:
I think it is because we need to update the bucket in the hash table 😔

Reply:
😥 Yes, the alternative is to store the bucket as I tried in previous.

But we need to reserve to ensure the hashtable big enough at the beginning of `intern`.

Otherwise when reserve and rehash happend, the bucket(actually a pointer to hashtable slot) will not point to the expected `(hash, group_index_view)` anymore.

I think it actually a bit danger to do so, and give up the try in later.